dbms/src/TiDB/Schema/SchemaBuilder.cpp

```diff
@@ -373,68 +391,100 @@ void SchemaBuilder<Getter, NameMapper>::applySetTiFlashReplica(DatabaseID databa
+    // Replica info is set to non-zero, create the storage if not exists.
+    auto storage = tmt_context.getStorages().get(keyspace_id, table_info->id);
+    if (storage == nullptr)
     {
-        // if set not 0, we first check whether the storage exists, and then check the replica_count and available
-        auto & tmt_context = context.getTMTContext();
-        auto storage = tmt_context.getStorages().get(keyspace_id, table_info->id);
-        if (storage != nullptr)
+        if (!table_id_map.tableIDInDatabaseIdMap(table_id))
         {
-            if (storage->getTombstone() == 0)
-            {
-                auto managed_storage = std::dynamic_pointer_cast<IManageableStorage>(storage);
-                auto storage_replica_info = managed_storage->getTableInfo().replica_info;
-                if (storage_replica_info.count == table_info->replica_info.count
-                    && storage_replica_info.available == table_info->replica_info.available)
-                {
-                    return;
-                }
-                else
-                {
-                    if (table_info->isLogicalPartitionTable())
-                    {
-                        for (const auto & part_def : table_info->partition.definitions)
-                        {
-                            auto new_part_table_info = table_info->producePartitionTableInfo(part_def.id, name_mapper);
-                            auto part_storage = tmt_context.getStorages().get(keyspace_id, new_part_table_info->id);
-                            if (part_storage != nullptr)
-                            {
-                                auto alter_lock = part_storage->lockForAlter(getThreadNameAndID());
-                                part_storage->alterSchemaChange(
-                                    alter_lock,
-                                    *new_part_table_info,
-                                    name_mapper.mapDatabaseName(db_info->id, keyspace_id),
-                                    name_mapper.mapTableName(*new_part_table_info),
-                                    context);
-                            }
-                            else
-                                table_id_map.emplacePartitionTableID(part_def.id, table_id);
-                        }
-                    }
-                    auto alter_lock = storage->lockForAlter(getThreadNameAndID());
-                    storage->alterSchemaChange(
-                        alter_lock,
-                        *table_info,
-                        name_mapper.mapDatabaseName(db_info->id, keyspace_id),
-                        name_mapper.mapTableName(*table_info),
-                        context);
-                }
-                return;
-            }
-            else
-            {
-                applyRecoverTable(db_info->id, table_id);
-            }
+            applyCreateTable(db_info->id, table_id);
         }
-        else
+        return;
+    }
```

Comment:
```
1. create table t (I int);
2. alter table t partition by hash(id) partitions 8; <-- normal table to partition table
3. alter tiflash replica t set tiflash replica 1;
4. insert into t values (10001), (10002), ....; <-- insert data
```

What if the ordering come to tiflash is `3->1->2->4`?

Reply:
Does `3->1` mean the "Region snapshot" comes before the DDL `create table t` is executed in TiFlash?

Reply:
yeah

1. tikv create normal table
2. tikv change the table to partition table
3. tikv send Region snapshot to tiflash
4. tiflash get `create table` DDL.

Reply:
When TiFlash try do decode the Region snapshot/raft command, it calls `AtomicGetStorageSchema`. Inside that function, if the storage instance is not yet created, it calls `syncTableSchema`. If `syncTableSchema` does not known the table id, it sync the DDL with tikv and apply `create table` in TiFlash.

https://github.com/pingcap/tiflash/blob/cf0e9affe7191b428a25142a91f240bce42823a1/dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp#L465-L488